File: app/schemas/test_run.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.test_run import TestRunStatus
from app.schemas.prompt import PromptRead, PromptVersionRead
from app.schemas.result import ResultRead


class TestRunBase(BaseModel):
    __test__ = False
    model_name: str = Field(..., max_length=100)
    model_version: str | None = Field(default=None, max_length=50)
    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    top_p: float = Field(default=1.0, ge=0.0, le=1.0)
    repetitions: int = Field(default=1, ge=1, le=50)
    schema_data: dict | None = Field(
        default=None, alias="schema", serialization_alias="schema"
    )
    notes: str | None = None
    batch_id: str | None = Field(default=None, max_length=64)

    model_config = ConfigDict(populate_by_name=True, serialize_by_alias=True)


class TestRunCreate(TestRunBase):
    __test__ = False
    prompt_version_id: int = Field(..., ge=1)


class TestRunUpdate(BaseModel):
    __test__ = False
    model_name: str | None = Field(default=None, max_length=100)
    model_version: str | None = Field(default=None, max_length=50)
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    top_p: float | None = Field(default=None, ge=0.0, le=1.0)
    repetitions: int | None = Field(default=None, ge=1, le=50)
    schema_data: dict | None = Field(
        default=None, alias="schema", serialization_alias="schema"
    )
    notes: str | None = None
    status: TestRunStatus | None = None
    batch_id: str | None = Field(default=None, max_length=64)

    model_config = ConfigDict(populate_by_name=True, serialize_by_alias=True)


class TestRunRead(TestRunBase):
    __test__ = False
    id: int
    prompt_version_id: int
    status: TestRunStatus
    created_at: datetime
    updated_at: datetime
    prompt_version: PromptVersionRead | None = None
    prompt: PromptRead | None = None
    results: list[ResultRead] = []

    model_config = ConfigDict(
        from_attributes=True, populate_by_name=True, serialize_by_alias=True
    )

File: frontend/src/types/testRun.ts
export type TestRunStatus = 'pending' | 'running' | 'completed' | 'failed'

export interface Metric {
  id: number
  result_id: number
  is_valid_json: boolean | null
  schema_pass: boolean | null
  missing_fields: Record<string, unknown> | null
  type_mismatches: Record<string, unknown> | null
  consistency_score: number | null
  numeric_accuracy: number | null
  boolean_accuracy: number | null
  created_at: string
}

export interface TestResult {
  id: number
  test_run_id: number
  run_index: number
  output: string
  parsed_output: Record<string, unknown> | null
  tokens_used: number | null
  latency_ms: number | null
  created_at: string
  metrics: Metric[]
}

export interface TestRun {
  id: number
  prompt_version_id: number
  batch_id: string | null
  model_name: string
  model_version: string | null
  temperature: number
  top_p: number
  repetitions: number
  schema: Record<string, unknown> | null
  status: TestRunStatus
  failure_reason: string | null
  notes: string | null
  created_at: string
  updated_at: string
  prompt_version?: import('./prompt').PromptVersion | null
  prompt?: import('./prompt').Prompt | null
  results: TestResult[]
}

export interface TestRunCreatePayload {
  prompt_version_id: number
  model_name: string
  model_version?: string | null
  temperature?: number
  top_p?: number
  repetitions?: number
  schema?: Record<string, unknown> | null
  notes?: string | null
  batch_id?: string | null
}

export interface ListTestRunParams {
  status?: TestRunStatus
  prompt_version_id?: number
  limit?: number
  offset?: number
}
